feat(isNaN): isNumberNaN

PatrickJS opened the pull request:
window.isNaN(‘lol’); //=> true
Number.isNaN(‘lol’); //=> false

isNaN converts it’s arguments into a Number before checking if it’s NaN. Number.isNaN fixes this but we can also polyfill it

pkozlowski-opensource commented:
@gdi2290 is there any reason to polyfill this particular [ES6 method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isNaN) inside the core? Shouldn't we rather advice people to include a full / partial polyfill if they need it?

gdi2290 commented:
well I didn't keep this global on the angular object so it's for internal use since it's inside of the closure. Throughout the code base the team is checking if it is a number before running isNaN. Adding this would save a few lines and from this js gotcha 

```javascript
if (isNumber(ctrl.$modelValue) && isNaN(ctrl.$modelValue)) {
// would be
if (isNaN(ctrl.$modelValue) {
//
minVal = isNumber(val) && !isNaN(val) ? val : undefined;
// would be
minVal = !isNaN(val) ? val : undefined;
```

pkozlowski-opensource commented:
@gdi2290 so, if you believe that the changed isNaN version can be used to improve existing code, could you please include those changes in this PR?

gdi2290 commented:
I think would be better if I name it `isNumberNaN` and replace all `isNaN` to `isNumberNaN` I'll update the PR

jbedard commented:
and instead of defining a function that wraps both the `Number.isNaN` and pollyfill it should only provide a pollyfill in place of `Number.isNaN`. Something such as...
```javascript
var isNumberNaN = Number.isNaN || function(n) {
   ...
};
```

gdi2290 commented:
good idea
```javascript
var _isNaN = window.isNaN;
var isNumberNaN = Number.isNaN || function isNumberNaN(num) {
  return isNumber(num) ? _isNaN(num) : false;
};
```

realityking commented:
Nice idea.

`Number.isNaN` can be more easily (and a bit faster) polyfilled like this

```javascript
function isNumberNaN(value) {
  return value !== value;
}
```

jbedard commented on src/ng/filter/limitTo.js:
```diff
@@ -96,12 +96,12 @@ function limitToFilter() {
     if (!isArray(input) && !isString(input)) return input;
 
-    begin = (!begin || isNaN(begin)) ? 0 : toInt(begin);
+    begin = (!begin || isNumberNaN(begin)) ? 0 : toInt(begin);
```
I don't think this one should change. This one specifically wanted `"123"` to pass...

jbedard commented on src/Angular.js:
```diff
@@ -399,6 +399,11 @@ function toInt(str) {
 
+var _isNaN = window.isNaN;
+var isNumberNaN = Number.isNaN || function isNumberNaN(num) {
+  return isNumber(num) ? _isNaN(num) : false;
```
This could be just `isNumber(num) && _isNaN(num)`. Or what @realityking said probably even better...

PatrickJS committed:
feat(isNaN): isNumberNaN

window.isNaN(‘lol’); //=> true
Number.isNaN(‘lol’); //=> false

isNaN converts it’s arguments into a Number before checking if it’s NaN

caitp commented on src/ng/directive/input.js:
```diff
@@ -1259,7 +1259,7 @@ function numberInputType(scope, element, attr, ctrl, $sniffer, $browser) {
         val = parseFloat(val, 10);
       }
-      minVal = isNumber(val) && !isNaN(val) ? val : undefined;
+      minVal = !isNumberNaN(val) ? val : undefined;
```
this changes semantic behaviour of the code
